src/storage-controller/src/lib.rs

```diff
@@ -3762,74 +3419,52 @@ where
-    pub write_frontier: Antichain<T>,
+impl<T: TimelyTimestamp> CollectionState<T> {
+    /// Returns true if `self` is a table, false otherwise.
+    pub fn is_table(&self) -> bool {
+        matches!(
+            self.data_source,
+            DataSource::Other(DataSourceOther::TableWrites)
+        )
+    }
+}
 
-    pub collection_metadata: CollectionMetadata,
+/// Additional state that the controller maintains for select collection types.
+#[derive(Debug)]
+enum CollectionStateExtra<T: TimelyTimestamp> {
+    Ingestion(IngestionState<T>),
+    None,
 }
 
-impl<T: Timestamp> CollectionState<T> {
-    /// Creates a new collection state, with an initial read policy valid from `since`.
-    pub fn new(
-        description: CollectionDescription<T>,
-        since: Antichain<T>,
-        write_frontier: Antichain<T>,
-        storage_dependency: Option<GlobalId>,
-        metadata: CollectionMetadata,
-    ) -> Self {
-        let mut read_capabilities = MutableAntichain::new();
-        read_capabilities.update_iter(since.iter().map(|time| (time.clone(), 1)));
-        Self {
-            description,
-            read_capabilities,
-            implied_capability: since.clone(),
-            read_policy: ReadPolicy::NoPolicy {
-                initial_since: since,
-            },
-            storage_dependency,
-            write_frontier,
-            collection_metadata: metadata,
-        }
-    }
+/// State maintained about ingestions and ingestion exports
+#[derive(Debug)]
+struct IngestionState<T: TimelyTimestamp> {
+    /// Really only for keeping track of changes to the `derived_since`.
+    pub read_capabilities: MutableAntichain<T>,
 
-    /// Returns the cluster to which the collection is bound, if applicable.
-    pub fn cluster_id(&self) -> Option<StorageInstanceId> {
-        match &self.description.data_source {
-            DataSource::Ingestion(ingestion) => Some(ingestion.instance_id),
-            DataSource::Webhook
-            | DataSource::Introspection(_)
-            | DataSource::Other(_)
-            // TODO(#8185) This isn't quite right because a source export runs
-            // on the ingestion's cluster, but we don't yet support announcing
-            // that.
-            | DataSource::IngestionExport { .. }
-            | DataSource::Progress => None,
-        }
-    }
+    /// The current since frontier, derived from `write_frontier` using
+    /// `hold_policy`.
+    pub derived_since: Antichain<T>,
 
-    /// Returns whether the collection was dropped.
-    pub fn is_dropped(&self) -> bool {
-        self.read_capabilities.is_empty()
-    }
+    /// Holds that this ingestion (or ingestion export) has on its dependencies.
+    pub dependency_read_holds: Vec<ReadHold<T>>,
+
+    /// Reported write frontier.
+    pub write_frontier: Antichain<T>,
+
+    /// The policy that drives how we downgrade our read hold. That is how we
+    /// derive our since from our upper.
+    pub hold_policy: ReadPolicy<T>,
```

Comment:
could we replace this (and all the `derive_since` and `read_capabilities` stuff) if we had a way to add a read policy to the policies set externally? Not a blocker, but definitely a bit weird that hand-manage a policy just to hold some `ReadHold`'s correctly

Reply:
consider me convinced! might be worth in the comment saying: `This is a _storage-controller-internal_ policy used to derive its personal read hold on the collection.`